samples/tensorRTEp/tensorrt_execution_provider.cc

```diff
@@ -0,0 +1,3770 @@
+    return nodes_list_output;
+  }
+
+  iterations++;
+  const size_t* node_index = nullptr;
+  size_t nodes_count = 0;
+  graph_api_->OrtGraph_GetNodesIndexInTopologicalOrder(graph, 1, &node_index, &nodes_count);
+  for (const auto& group : nodes_vector_input) {
+    // Construct subgraph
+    if (!group.first.empty()) {
+      if (group.second) {
+        nodes_list_output.push_back(group);
+      } else {
+
+        const OrtGraphViewer* sub_graph_viewer = nullptr;
+        graph_api_->OrtGraph_GetSubGraph(graph, group.first.size(), group.first.data(), &sub_graph_viewer);
+
+        void* buf_data = nullptr;
+        size_t buf_size = 0;
+        graph_api_->OrtGraph_SerializeToArray(sub_graph_viewer, &buf_data, &buf_size);
+
+        // Get supported node list recursively
+        SubGraphCollection_t parser_nodes_list;
+        TensorrtLogger& trt_logger = GetTensorrtLogger(detailed_build_log_);
+        auto trt_builder = GetBuilder(trt_logger);
+        auto network_flags = 0;
+#if NV_TENSORRT_MAJOR > 8
+        network_flags |= fp16_enable_ || int8_enable_ ? 0 : 1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kSTRONGLY_TYPED);
+#endif
+        network_flags |= 1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
+        auto trt_network = std::unique_ptr<nvinfer1::INetworkDefinition>(trt_builder->createNetworkV2(network_flags));
+
+        auto trt_parser = tensorrt_ptr::unique_pointer<nvonnxparser::IParser>(nvonnxparser::createParser(*trt_network, trt_logger));
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4996)
+#endif
+        trt_parser->supportsModel(buf_data, buf_size, parser_nodes_list, model_path_);
+        graph_api_->OrtFreeMem(buf_data);
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+
+        SubGraphCollection_t next_nodes_list;
+        const size_t* subgraph_node_index = nullptr;
+        size_t subgraph_node_count = 0;
+        graph_api_->OrtGraph_GetNodesIndexInTopologicalOrder(sub_graph_viewer, 1, &subgraph_node_index, &subgraph_node_count);
+        next_nodes_list = GetSupportedList(parser_nodes_list, iterations, max_iterations, sub_graph_viewer, early_termination);
+        for (size_t i = 0, end = next_nodes_list.size(); i < end; ++i) {
+          for (size_t j = 0, end = next_nodes_list[i].first.size(); j < end; ++j) {
+            next_nodes_list[i].first[j] = group.first[subgraph_node_index[next_nodes_list[i].first[j]]];
+          }
+          nodes_list_output.push_back(next_nodes_list[i]);
+        }
+        graph_api_->OrtGraph_ReleaseGraphViewer(sub_graph_viewer);
+      }
+    }
+  }
+  return nodes_list_output;
+}
+
+}   // namespace onnxruntime
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+OrtExecutionProviderFactory* RegisterCustomEp() {
+    std::unique_ptr<onnxruntime::TensorrtExecutionProviderFactory> ret = std::make_unique<onnxruntime::TensorrtExecutionProviderFactory>();
+    return ret.release();
+}
```

Comment:
After more investigation, it seems like this implementation directly returns a `OrtExecutionProviderFactory*` because you want to be able to inherit from `OrtExecutionProviderFactory` and use polymorphism for calls to `GetCapability` and `Compile`.

Here's an alternative approach that keeps allocations on the EP side of the API boundary and still allows you to create a custom TensorRTEP class.
```C++
// tensorrt_execution_provider.cc (EP Plugin Library).
#include <onnxruntime_c_api_ep.h>
struct TensorRTEpFactory {
 // …
 std::string ep_name;
 OrtExecutionProviderFactory ort_callbacks = {};
};

class TensorRTEp {
public:
 static std::unique_ptr<TensorRTEp> CreateInstance(const std::string& ep_name, size_t num_ep_options,
                                                   const char* const* ep_option_keys, const char* const* ep_option_vals,
                                                   /*output*/ std::string& err_msg) {
   if (/* EP options invalid */) { err_msg = "Invalid EP options"; return nullptr; }

   // Create internal EP object/state and fill out ORT callbacks (only GetCapability and Compile are shown here).
   std::unique_ptr<TensorRTEp> ep = std::make_unique<TensorRTEp>(ep_name, /*other args*/);
   ep->ort_ep.GetCapability = [](const OrtExecutionProvider* ort_ep, /*params*/) {
     TensorRTEp* this_ = reinterpret_cast<TensorRTEp*>(ort_ep->state);
     /* impl */
   };
   ep->ort_ep.Compile = [](const OrtExecutionProvider* ort_ep, /*params*/) {/* impl */};
   ep->ort_ep.state = ep->get();
   return ep;
 }

 OrtExecutionProvider* GetOrtExecutionProvider() { return &this->ort_ep; }
private:
 TensorRTEp(const std::string& ep_name, /* other params */) { /* … */ }

 std::string ep_name;
 OrtExecutionProvider ort_ep = {};
 // Other state/methods here …
};

static void DestroyOnDllUnload(std::unique_prt<TensorRTEpFactory>&& ep_factory) {
  static std::vector<std::unique_prt<TensorRTEp>> ep_factories;
  static std::mutex m_;
  std::lock_guard<std::mutex> lock(m_);
  ep_factories.push_back(std::move(ep_factory));
}
static void DestroyOnDllUnload(std::unique<TensorRTEp> ep_instance) {/*similar implementation as above*/}

#ifdef __cplusplus
extern "C" {
#endif
// DLL ENTRY POINT
OrtStatus* ORT_API_CALL RegisterCustomEp(const OrtExecutionProviderFactory** ort_ep_factory, const char* ep_name) {
    auto ep_factory = std::make_unique<TensorRTEpFactory>(ep_name, OrtExecutionProviderFactory{});
    ep_factory->ort_callbacks.CreateExecutionProvider = CreateExecutionProviderCallback;
    ep_factory->ort_callbacks.state = ep_factor->get(); // Can optionally store custom state.
    
    *ort_ep_factory = &ep_factory->ort_ep_factory; // Update output parameter to point to our EP factory callbacks. 

    DestroyOnDllUnload(std::move(ep_factory));
    return nullptr;
}
#ifdef __cplusplus
}
#endif

OrtStatus* ORT_API_CALL CreateExecutionProviderCallback(const OrtExectutionProviderFactory* ort_ep_factory,
                                                        const char* const* option_keys,
                                                        const char* const* option_vals,
                                                        size_t num_options,
                                                        const OrtExecutionProvider** ort_ep) {
  TensorRTEpFactory* ep_factory = reinterpret_cast<TensorRTEpFactory*>(ort_ep_factory->state);
  std::string err_msg;
  auto ep = TensorRTEp::CreateInstance(ep_factory->ep_name, option_keys, option_vals, num_options, err_msg);

  if (!err_msg.empty()) { /* return OrtStatus with error message */ }
  *ort_ep = ep->GetOrtExecutionProvider(); // Update output parameter to point to our EP callbacks.
  DestroyOnDllUnload(std::move(ep));
  return nullptr;
}
```

The above requires modifying `OrtExecutionProviderFactory` and `OrtExecutionProvider` to store a `void*` pointer to custom state.

```C++
struct OrtExecutionProviderFactory {
 // Same
 void* state;  // State set by the EP plugin library.
};

struct OrtExecutionProvider {
 // Same
 void* state;  // State set by the EP plugin library.
};
```

Reply:
There's another alternative. In the above example, ORT receives pointers to `OrtExecutionProviderFactory` and `OrtExecutionProvider` structs that were allocated within the EP plugin library. An alternative is to just copy the structs since they’re bags of function pointers (plain-old-data). There is really no need to give ORT a pointer to these structs when we can just copy them so that ORT has its own versions. This would be achievable with the following changes:

```C++
#include <onnxruntime_c_api_ep.h>

// Same as previous example

#ifdef __cplusplus
extern "C" {
#endif
// DLL ENTRY POINT
OrtStatus* ORT_API_CALL RegisterCustomEp(/*out*/ OrtExecutionProviderFactory* ort_ep_factory, const char* ep_name) {
    auto ep_factory = std::make_unique<TensorRTEpFactory>(ep_name, OrtExecutionProviderFactory{});
    ep_factory->ort_ep_factory.CreateExecutionProvider = CreateExecutionProviderCallback;
    ep_factory->ort_ep_factory.state = ep_factor->get(); // Can optionally store custom state.
    
    // This is a struct copy. This updates the output parameter to a copy of our EP factory callbacks.
    *ort_ep_factory = ep_factory->ort_ep_factory;
    DestroyOnDllUnload(std::move(ep_factory));
    return nullptr;
}
#ifdef __cplusplus
}
#endif

OrtStatus* ORT_API_CALL CreateExecutionProviderCallback(const OrtExectutionProviderFactory* ort_ep_factory,
                                                        const char* const* option_keys,
                                                        const char* const* option_vals,
                                                        size_t num_options,
                                                        /*output*/ OrtExecutionProvider* ort_ep) {
  TensorRTEpFactory* ep_factory = reinterpret_cast<TensorRTEpFactory*>(ort_ep_factory->state);
  std::string err_msg;
  auto ep = TensorRTEp::CreateInstance(ep_factory->ep_name, option_keys, option_vals, num_options, err_msg);

  if (!err_msg.empty()) { /* return OrtStatus with error message */ }
  *ort_ep = *(ep->GetOrtExecutionProvider()); // Struct copy. This updates the out param to a copy of our EP callbacks.
  DestroyOnDllUnload(std::move(ep));
  return nullptr;
}
```